PBR/src/PBR_Renderer.cpp

```diff
@@ -480,21 +480,45 @@ void PBR_Renderer::PrecomputeCubemaps(IDeviceContext* pCtx,
         float4x4 Rotation;
 
         float Roughness;
-        float EnvMapDim;
+        float EnvMapWidth;
+        float EnvMapHeight;
         uint  NumSamples;
-        float Dummy;
+    };
+
+    auto GetIBL_PSO = [&](IBL_FEATURE_FLAGS FeatureFlags, TEXTURE_FORMAT RTVFormat) -> RefCntAutoPtr<IPipelineState>& {
+        auto Iter = m_IBL_PSOCache.find(IBL_PSOKey{FeatureFlags, RTVFormat});
+        if (Iter != m_IBL_PSOCache.end())
+            return Iter->second;
+
+        auto Condition = m_IBL_PSOCache.emplace(IBL_PSOKey{FeatureFlags, RTVFormat}, RefCntAutoPtr<IPipelineState>{});
+        return Condition.first->second;
     };
 
     if (!m_PrecomputeEnvMapAttribsCB)
     {
         CreateUniformBuffer(m_Device, sizeof(PrecomputeEnvMapAttribs), "Precompute env map attribs CB", &m_PrecomputeEnvMapAttribsCB);
     }
 
-    if (!m_pPrecomputeIrradianceCubePSO)
+    IBL_FEATURE_FLAGS FeatureFlags = IBL_FEATURE_FLAG_NONE;
+
+    if (OptimizeSamples)
+        FeatureFlags = static_cast<IBL_FEATURE_FLAGS>(FeatureFlags | IBL_FEATURE_FLAG_OPTIMIZE_SAMPLES);
+
+    if (!pEnvironmentMap->GetTexture()->GetDesc().IsCube())
+        FeatureFlags = static_cast<IBL_FEATURE_FLAGS>(FeatureFlags | IBL_FEATURE_FLAG_SAMPLING_SPHERE);
+
+    auto pPrecomputeIrradianceCubePSO = GetIBL_PSO(FeatureFlags, m_pIrradianceCubeSRV->GetDesc().Format);
```

Comment:
Here is a bug: the type of pPrecomputeIrradianceCubePSO will not be reference, and the PSO will never be written to cache